File: services/frontend/app/static/js/dashboard/ports.js
// ports.js renderer switch-port oversigten i dashboardet.
// Data kommer fra dashboardData.ports, som frontend modtager via /api/live-dashboard.
// Backend bygger portdata ved at kombinere SNMP-data fra switch_monitor.py med devices og events.
// Den viser kun porte, portstatus og de devices backend har mappet til hver port.


// Dataflow:
// switch_monitor.py + backend dashboard/ports.py
// └─ backend /api/dashboard
//    └─ dashboardData.ports
//       └─ renderPorts(container, dashboardData)
//          ├─ bygger ét port-card pr. switch-port
//          └─ renderDevice(device) viser device-info under porten
import { escapeHtml } from "./utils.js";

// renderDevice() bygger HTML for én device der er mappet til en switch-port.
// Device-data er allerede koblet til porten af backend.
function renderDevice(device) {
    // unit_ids vises kun hvis backend har fundet Modbus unit IDs for devicen.
    const unitText = (device.unit_ids && device.unit_ids.length)
        ? ` | unit ${device.unit_ids.join(", ")}`
        : "";

    // Returnerer HTML-kortet for devicen inde i port-cardet.
    return `
        <div class="port-device">
            <div class="port-device-top">
                <strong>${device.label || "Device"}</strong>
                <span class="badge ${device.role || "inactive"}">${device.role || "unknown"}</span>
            </div>
            <p>${device.ip || "-"}</p>
            <p>${device.mac || "-"}${unitText}</p>
        </div>
    `;
}

// renderPorts() tegner hele switch-port sektionen.
// container er HTML-elementet hvor port-kortene indsættes.
// dashboardData indeholder ports-listen fra backend.
export function renderPorts(container, dashboardData) {
    // Hvis backend ikke sender ports, bruges en tom liste så renderingen ikke fejler.
    const ports = dashboardData.ports || [];

    // Hvis der ikke er portdata, vises en tom-tilstand i dashboardet.
    if (!ports.length) {
        container.innerHTML = `<div class="chart-empty-state">No switch port data available.</div>`;
        return;
    }

    // Bygger ét synligt port-card pr. switch-port.
    container.innerHTML = ports.map((port) => {
        // devices er de enheder backend har mappet til denne fysiske port.
        const devices = port.devices || [];
        // Hvis der ikke er mapped devices, vises en neutral tom-tekst under porten.
        const devicesHtml = devices.length
            ? `<div class="port-devices">${devices.map(renderDevice).join("")}</div>`
            : `<div class="port-devices-empty">No mapped device</div>`;

        // Port-cardet viser portnavn, fysisk/interface-navn, state, speed, activity og mapped devices.
        return `
            <div class="port-card ${port.state || "inactive"}">
                <div class="port-head">
                    <div>
                        <h4>${port.port}</h4>
                        <p>${port.name || "-"}</p>
                    </div>
                    <span class="badge ${port.state || "inactive"}">${port.state || "inactive"}</span>
                </div>
                <p><strong>Speed:</strong> ${port.speed || "-"}</p>
                <p><strong>Activity:</strong> ${port.activity || "-"}</p>
                ${devicesHtml}
            </div>
        `;
    }).join("");
}

File: services/backend/app/dashboard/ports.py
# ports.py bygger port-sektionen til dashboardet.
# Filen kombinerer data fra switch_monitor, devices-tabellen og observed_connections.
# Resultatet er en liste over switch-porte med status, tilknyttede devices, VLANs og event_id til port-alarmer.
# Filen læser switch-data via SNMP-funktioner i switch_monitor.py og bruger storage-writeren til at oprette port_active events.
from switch_monitor import get_mac_to_port_map, get_switch_ports
from storage import get_writer


# attach_port_events() opretter et event for hver aktiv switch-port.
# Kun porte hvor state == "active" får et port_active event.
# event_key bygges som port_active:<portnavn>, så samme aktive port genbruger samme event i events-tabellen.
# event_id gemmes direkte på port-dictet, så frontend kan sende alarm_approvals.event_id tilbage ved godkendelse.
def attach_port_events(ports):
    writer = get_writer()

    for port in ports:
        # state læses fra port-data og gøres lowercase, så sammenligningen med "active" er stabil.
        state = str(port.get("state") or "").lower()
        if state != "active":
            continue

        # port_name vælger den bedste tilgængelige port-identitet.
        # Den bruges i event_key, så samme port kan genkendes næste gang dashboardet opdateres.
        port_name = port.get("port") or port.get("name") or port.get("ifname") or "unknown"
        event_id = writer.insert_event(
            event_key=f"port_active:{port_name}",
            event_type="port_active",
            severity="medium",
            details={
                "message": "Switch port is active",
                "port": port.get("port"),
                "name": port.get("name"),
                "state": port.get("state"),
                "activity": port.get("activity"),
            },
        )
        port["event_id"] = event_id

    return ports


# build_connection_groups() samler observed_connections-rækker efter master_ip.
# Input kommer fra dashboard/queries.py get_connection_rows().
# Output bruges af frontend til at vise hvilke slaves hver master kommunikerer med.
# unit_id vises sammen med slave_ip, fordi samme slave-IP kan have flere Modbus unit IDs.
def build_connection_groups(rows): 
    groups = {}

    for row in rows:
        master = row["master_ip"]
        # Første gang en master ses, oprettes en gruppe til dens slaves.
        if master not in groups:
            groups[master] = {
                "master": master,
                "slave_count": 0,
                "last_seen": row["last_seen"],
                "slaves": [],
            }

        # Hver database-række repræsenterer én master -> slave -> unit_id relation.
        groups[master]["slaves"].append(
            {
                "ip": f"{row['slave_ip']} (unit {row['unit_id']})" if row["unit_id"] else row["slave_ip"],
                "status": "online",
                "packets": row["request_count"],
                "last_seen": row["last_seen"],
            }
        )
        groups[master]["slave_count"] += 1
        groups[master]["last_seen"] = max(groups[master]["last_seen"], row["last_seen"])

    return list(groups.values())


# enrich_ports_with_devices() kobler switch-porte sammen med kendte devices.
# get_mac_to_port_map() henter switchens MAC->port map fra switch_monitor.py.
# devices kommer fra devices-tabellen, og connections kommer fra observed_connections.
# Formålet er at vise hvilke IP/MAC-adresser der sidder på hvilken fysisk switch-port.
def enrich_ports_with_devices(ports, devices, connections):
    mac_to_port = get_mac_to_port_map()

    # device_by_ip gør det hurtigt at slå en device op på IP uden CIDR-suffix.
    # split("/")[0] fjerner f.eks. /32, hvis PostgreSQL INET returnerer IP med suffix.
    device_by_ip = {
        (device.get("ip") or "").split("/")[0]: device
        for device in devices
        if device.get("ip")
    }

    # connection_map bruges til at give en device en rolle ud fra Modbus-kommunikationen.
    # Hvis en IP står som master i observed_connections, får den role_hint master.
    # Hvis en IP står som slave, gemmes dens unit_ids.
    connection_map = {}

    for group in connections:
        master_ip = (group.get("master") or "").split("/")[0]
        connection_map.setdefault(master_ip, {"role_hint": "master", "unit_ids": set()})

        for slave in group.get("slaves", []):
            raw_ip = (slave.get("ip") or "").split(" ")[0]
            slave_ip = raw_ip.split("/")[0]
            unit_text = slave.get("ip") or ""
            unit_id = None

            if "(unit " in unit_text:
                try:
                    unit_id = int(unit_text.split("(unit ")[1].split(")")[0])
                except Exception:
                    unit_id = None

            entry = connection_map.setdefault(slave_ip, {"role_hint": "slave", "unit_ids": set()})
            if unit_id is not None:
                entry["unit_ids"].add(unit_id)

    # port_index gør det hurtigt at finde port-dictet ud fra portnavnet fra switchens MAC-table.
    port_index = {port["port"]: port for port in ports}

    # Her kobles devices til switch-porte.
    # Device MAC bruges til at slå porten op i mac_to_port fra switchen.
    for ip, device in device_by_ip.items():
        mac = (device.get("mac") or "").lower()
        if not mac:
            continue

        mapping = mac_to_port.get(mac)
        if not mapping:
            continue

        port = port_index.get(mapping["port"])
        if not port:
            continue

        conn_info = connection_map.get(ip, {})
        role = device.get("role") or conn_info.get("role_hint") or "unknown"
        unit_ids = sorted(conn_info.get("unit_ids", set()))
        # label er et frontend-navn baseret på rollen: slave vises som PLC, master som Master.
        label = "PLC" if role == "slave" else "Master" if role == "master" else "Device"

        port["devices"].append(
            {
                "ip": ip,
                "mac": mac,
                "role": role,
                "label": label,
                "unit_ids": unit_ids,
                "vlan_id": mapping.get("vlan_id"),
                "ifname": mapping.get("ifname"),
                "ifindex": mapping.get("ifindex"),
            }
        )

    # Efter devices er koblet på porte, beregnes VLANs og duplicate MAC alarms pr. port.
    for port in ports:
        vlan_ids = sorted({
            device["vlan_id"]
            for device in port.get("devices", [])
            if device.get("vlan_id") is not None
        })
        port["vlans"] = vlan_ids

        mac_groups = {}
        for device in port.get("devices", []):
            mac = (device.get("mac") or "").lower()
            ip = device.get("ip")
            if not mac or not ip:
                continue
            mac_groups.setdefault(mac, []).append(ip)

        # duplicate_macs viser om samme MAC er knyttet til flere IP'er på samme port.
        # Det kan indikere uventet netværksadfærd og markeres til frontend.
        duplicate_macs = [
            {"mac": mac, "ips": sorted(set(ips))}
            for mac, ips in mac_groups.items()
            if len(set(ips)) > 1
        ]

        port["duplicate_macs"] = duplicate_macs
        port["has_duplicate_mac_alarm"] = bool(duplicate_macs)

    return ports


# build_ports() er hovedfunktionen som dashboard/service.py bruger.
# Først hentes portstatus fra switch_monitor.py.
# Derefter får aktive porte event_id via attach_port_events().
# Til sidst kobles devices og Modbus connections på portene med enrich_ports_with_devices().
def build_ports(devices, connections):
    ports = get_switch_ports()
    ports = attach_port_events(ports)
    return enrich_ports_with_devices(ports, devices, connections)

